src/internal/storage/chunk/batcher.go

```diff
@@ -0,0 +1,133 @@
+type ChunkFunc = func([]interface{}, *DataRef) error
+
+// EntryFunc is a function that provides the metadata for an entry and a data reference to the entry in a chunk.
+// Size zero entries will have a nil data reference.
+type EntryFunc = func(interface{}, *DataRef) error
+
+// Batcher batches entries into chunks.
+// Entries are buffered until they are past the configured threshold, then a chunk is created.
+// Chunk creation is asynchronous with respect to the client, which is why the interface
+// is callback based.
+// Batcher provides one of two callback based interfaces defined by ChunkFunc and EntryFunc.
+// Callbacks will be executed with respect to the order the entries are added (for the ChunkFunc
+// interface, entries are ordered within as well as across calls).
+type Batcher struct {
+	client    Client
+	entries   []*entry
+	buf       []byte
+	threshold int
+	taskChain *TaskChain
+	chunkFunc ChunkFunc
+	entryFunc EntryFunc
+}
+
+type entry struct {
+	meta     interface{}
+	size     int
+	dataRef  *DataRef
+	pointsTo []*DataRef
+}
+
+// TODO: Add config for number of entries.
+func (s *Storage) NewBatcher(ctx context.Context, name string, threshold int, opts ...BatcherOption) *Batcher {
+	client := NewClient(s.store, s.db, s.tracker, NewRenewer(ctx, s.tracker, name, defaultChunkTTL))
+	b := &Batcher{
+		client:    client,
+		threshold: threshold,
+		taskChain: NewTaskChain(ctx, semaphore.NewWeighted(chunkParallelism)),
+	}
+	for _, opt := range opts {
+		opt(b)
+	}
+	return b
+}
+
+func (b *Batcher) Add(meta interface{}, data []byte, pointsTo []*DataRef) error {
+	b.entries = append(b.entries, &entry{
+		meta:     meta,
+		size:     len(data),
+		pointsTo: pointsTo,
+	})
+	b.buf = append(b.buf, data...)
+	if len(b.buf) >= b.threshold {
+		if err := b.createBatch(b.entries, b.buf); err != nil {
+			return err
+		}
+		b.entries = nil
+		b.buf = nil
+	}
+	return nil
+}
+
+func (b *Batcher) createBatch(entries []*entry, buf []byte) error {
+	return b.taskChain.CreateTask(func(ctx context.Context) (func() error, error) {
+		pointsTo := getPointsTo(entries)
+		dataRef, err := upload(ctx, b.client, buf, pointsTo, false)
+		if err != nil {
+			return nil, err
+		}
+		// Handle chunk callback.
+		if b.chunkFunc != nil {
+			var metas []interface{}
+			for _, entry := range entries {
+				metas = append(metas, entry.meta)
+			}
+			return func() error {
+				return b.chunkFunc(metas, dataRef)
+			}, nil
+		}
+		// Handle entry callback.
+		if b.entryFunc != nil {
+			offset := 0
+			for _, entry := range entries {
+				if entry.size == 0 {
+					continue
+				}
+				entry.dataRef = NewDataRef(dataRef, buf, int64(offset), int64(entry.size))
+				offset += entry.size
+			}
+			return func() error {
+				for _, entry := range entries {
+					if err := b.entryFunc(entry.meta, entry.dataRef); err != nil {
+						return err
+					}
+				}
+				return nil
+			}, nil
+		}
```

Comment:
Having the two options feels a little messy to me vs just returning `dataRef, metas, []computedDataRefs` and having the caller use what they want, but I don't feel too strongly about it

Reply:
I prefer this approach because it makes things simpler for the client and the additional complexity in this abstraction is minimal. Also, it is slightly more performant since the hashes aren't computed for the chunk case (index entries).